File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    nl=len(numbers)
    cnt=0
    def dfs(depth, arr):
        nonlocal nl,target,cnt
        if depth==nl:
            s=sum(arr)
            if s==target:
                cnt+=1
            return
        
        dfs(depth+1,arr+[numbers[depth]])
        dfs(depth+1,arr+[-numbers[depth]])
    
    # depth, arr
    dfs(0,[0])
    return cnt

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    public static int cnt=0;
    public static void dfs(int depth,int total,int[] numbers,int target){
        if(depth==numbers.length){
            if(total==target){
                cnt++;
            }
            return;
        }
        
        dfs(depth+1,total+numbers[depth],numbers,target);
        dfs(depth+1,total-numbers[depth],numbers,target);
    }
    
    public int solution(int[] numbers, int target) {
        dfs(0,0,numbers,target);
        return cnt;
    }
}